### ISSUE TITLE
HDR tonemapping not working on external display

### ISSUE THREAD

**Issue by burgerga** [iina/iina#5119]
**System and IINA version:**

- macOS 14.6 (23G80)
- IINA 1.3.5 Build 141 (mpv 0.35.0-419-gf79458476b-dirty, FFmpeg 6.0)

**Expected behavior:**
IINA applies tone mapping to my HDR video so I can watch it on my external display without HDR support

**Actual behavior:**
tone mapping doesn't seem to do anything at all (setting target peak to multiple options between 100-320)

See difference in behavior between IINA, mpv, and Infuse (look at the eyes of the left character):
![Screenshot 2024-08-08 at 15 29 42](https://github.com/user-attachments/assets/6dfafb85-1ca4-4c50-b83f-1c7887b00b87)
IINA is darkest, mpv is only slightly brighter, Infuse is brightest

Now when we're trying to tone map:
![Screenshot 2024-08-08 at 16 31 55](https://github.com/user-attachments/assets/ed9bec21-4835-437a-bbdb-b965564db718)

Now the strange thing is that when I drag the window to the Macbook Pro Display, the IINA log says:
```
16:38:05.535 [iina][d] Falling back to nominal display refresh rate: 120.0 from 0.0
16:38:05.536 [mpv0][d] Set property: override-display-fps=120.0
16:38:05.540 [iina][d] Refreshing HDR for player0 @ display1: "Built-in Retina Display" visible frame (-1684.0, 0.0, 1684.0, 1079.0) EDR: {supports=true maxPotential=15.989999771118164 maxCurrent=1.0}
16:38:05.540 [hdr][d] HDR gamma=pq, primaries=bt.2020, sig_peak=5.133005142211914
16:38:05.540 [hdr][d] Will activate HDR color space instead of using ICC profile
16:38:05.540 [mpv0][d] Set property: icc-profile=
16:38:05.540 [mpv0][d] Set property: target-prim=bt.2020
16:38:05.540 [mpv0][d] Set property: target-trc=pq
16:38:05.540 [mpv0][d] Set property: screenshot-tag-colorspace=true
16:38:05.541 [hdr][d] Will enable tone mapping target-peak=100 algorithm=auto
16:38:05.541 [mpv0][d] Set property: target-peak=100
16:38:05.541 [mpv0][d] Set property: tone-mapping=auto
```

But then when I draw it back to the external display it says:
```
16:40:03.187 [iina][d] Falling back to nominal display refresh rate: 49.98677433262449 from 0.0
16:40:03.187 [mpv0][d] Set property: override-display-fps=49.98677433262449
16:40:03.188 [iina][d] Refreshing HDR for player0 @ display2: "34UM95" visible frame (0.0, 0.0, 3440.0, 1415.0) EDR: {supports=false maxPotential=1.0 maxCurrent=1.0}
16:40:03.188 [hdr][d] HDR gamma=pq, primaries=bt.2020, sig_peak=5.133005142211914
16:40:03.188 [hdr][d] HDR video was found but the display does not support EDR mode
16:40:03.188 [hdr][d] Loading ICC profile
16:40:03.192 [mpv0][d] Set property: icc-profile=/Library/ColorSync/Profiles/Displays/34UM95-0FE29938-AFCB-4FB0-8DFB-F94D1D12DAFE.icc
16:40:03.192 [mpv0][d] Set property: target-trc=auto
16:40:03.192 [mpv0][d] Set property: target-prim=auto
16:40:03.192 [mpv0][d] Set property: target-peak=auto
16:40:03.192 [mpv0][d] Set property: tone-mapping=auto
16:40:03.192 [mpv0][d] Set property: tone-mapping-param=default
16:40:03.192 [mpv0][d] Set property: screenshot-tag-colorspace=false
16:40:03.877 [mpv0][d] Set property: window-scale=0.44765625

```

**seemingly ignoring the tone mapping settings...**
I've also attached the mpv log below

[mpv.log](https://github.com/user-attachments/files/16549890/mpv.log)

**Steps to reproduce:**

- [x] MPV does not have this problem.

**How often does this happen?**
Always

**Comment by low-batt** [comment]
I believe you are correct and tone mapping is only applied when HDR mode is active. I'm not sure why the decision was made to only apply it when HDR mode was active.

@CarterLi Can you comment on the situation with tone mapping for HDR videos played on a display that is not capable of EDR?

I believe the darkness compared to mpv is due to other IINA issues. I have been looking into those.

On mpv being darker than Infuse, the log file shows the video in question uses Dolby Vision. Better support for DV is being added to mpv in [gpu-next](https://mpv.io/manual/master/#video-output-drivers-gpu-next), which is still experimental. If the build of mpv you are using included support for [MoltenVK](https://moltengl.com/moltenvk/) then you may be able to try mpv with `vo=gpu-next` and see if that makes a difference.

Another difference is that the versions of libraries from [mpv](https://github.com/mpv-player/mpv) and [FFmpeg](https://ffmpeg.org/) IINA 1.3.5 is using are now a good bit out of date. IINA is currently under heavy development. The libraries have been updated to mpv 0.38.0 and FFmpeg 7.0.1 in the current development version of IINA.

**Comment by burgerga** [comment]
Thanks for your fast reply, those are some interesting points to look into! I will try to get a development version working.

Offtopic: HDR (and especially DV) have been hit or miss :) For some DV files I actually got better results in IINA (maybe because of HDR10 fallback?) than in Infuse, and Infuse only showed me "proper" HDR when I turned Dolby Vision off 😕

**Comment by low-batt** [comment]
If you want to test the development version of IINA, nightly builds are listed [here](https://iina.io/nightly/). You would want one from the `develop` branch. **NOTE** that nightly builds are not signed and as of Sonoma, macOS will clear the `Open Recent` list if code signatures do not match. Very irritating for developers. I've not been able to find people on the net talking about this. There is a workaround for this in IINA 1.3.5, but there are problems with that workaround if the list contains references to files in devices that were mounted and are not available when IINA starts. See issues #4688 and #4973 for more info.

If you notice a regression in the development version of IINA, or especially a hang/freeze, please report it. Lots of changes to address [technical debt](https://en.wikipedia.org/wiki/Technical_debt) and threading problems have already been merged into the `develop` branch.

Another aspect of mpv 0.38.0 I did not mention is that if it is built with [MoltenVK](https://moltengl.com/moltenvk/), even if [gpu-next](https://mpv.io/manual/master/#video-output-drivers-gpu-next) is not used it will default to using [libplacebo](https://github.com/haasn/libplacebo) and MoltenVK/Metal. To make mpv 0.38.0 revert back to OpenGL, use `--vo=libmpv`. I'm surprised that makes a difference in the picture displayed. Maybe improvements in libplacebo that do not require gpu-next?

**Comment by CarterLi** [comment]
> @CarterLi Can you comment on the situation with tone mapping for HDR videos played on a display that is not capable of EDR?

If you play HDR videos on a SDR display, all HDR stuff will be disabled, including tone mapping.

https://github.com/iina/iina/blob/develop/iina/VideoView.swift#L478-L481

MPV should convert it to SDR video when playing.

**Comment by burgerga** [comment]
Building mpv using homebrew (using https://github.com/pxia/mpv-macos-build) already gives better results:
![Screenshot 2024-08-13 at 12 14 31](https://github.com/user-attachments/assets/8a83f813-7777-4a73-9dd1-26bcc8849e9a)

I installed molten-vk as well, but weirdly enough it doesn't matter if I use `--vo=gpu-next` (libplacebo) or `--vo=libmpv`, the improvement in brightness is there in both cases...

@CarterLi Would there be a downside to passing on the tone mapping settings to mpv for SDR screens?

**Comment by burgerga** [comment]
I actually get very good results with `mpv --tone-mapping=bt.2446a` (using mpv 0.38 with gpu-next) which gives me very similar output to Infuse (look at the eyes):
![Screenshot 2024-08-13 at 14 30 01](https://github.com/user-attachments/assets/23cfea75-5ea5-497a-a97d-577876d7f9fc)

So I think it would be great if tone-mapping settings could be passed through to mpv for SDR screens as well, because, as you can see in my previous comment, the mpv default is still significantly darker.

**Comment by burgerga** [comment]
I thought I would do a quick-and-dirty edit in https://github.com/iina/iina/blob/develop/iina/VideoView.swift#L478-L481 to test it out, but alas, I got:
```
[mpv0][w] [libmpv_render] warn: Tone mapping curve unsupported by vo_gpu, falling back.
```

@low-batt I guess that means we have to wait for either https://github.com/mpv-player/mpv/pull/13620 or https://github.com/mpv-player/mpv/issues/10810?

**Comment by low-batt** [comment]
The mpv documentation for [tone-mapping](https://mpv.io/manual/stable/#options-tone-mapping) confirms your results:
> **bt.2446a**
    HDR<->SDR mapping specified in ITU-R Report BT.2446, method A. This is the recommended curve for well-mastered content. (`--vo=gpu-next` only)

As mpv is moving to `gpu-next` many new features are only being added for to that video driver. And yes, IINA has to wait for https://github.com/mpv-player/mpv/issues/10810 to be addressed before IINA can support `gpu-next`. Major bummer since there are a lot of improvements in color management specific to `gpu-next` listed [here](https://github.com/mpv-player/mpv/wiki/GPU-Next-vs-GPU#general-differences).

You should be able to use one of the other tone mapping algorithms that works for `--vo=gpu` although I expect bt.2446a is the one that is best for this.

I have the same question for @CarterLi as you asked: Why can't tone mapping always be used?

The current development version of IINA, which uses mpv 0.38.0 should match up with mpv 0.38.0 when using `--vo=libmpv`, but in some cases the picture does not match. The IINA view layer code needs to be updated. I am currently trying to finish up what I was working on. I am then going to turn my attention to this which is being tracked in issue #5124. I have already experimented with changes and was able to get IINA to match mpv  (with `--vo=libmpv`) in the test case I was using.

I believe some of those color management improvements listed for `gpu-next` come from [libplacebo](https://github.com/haasn/libplacebo). I think that is why mpv without `--vo=libmpv` looks better. Not sure. I've not been following this aspect of mpv development closely since  https://github.com/mpv-player/mpv/issues/10810 has not been addressed.

**Comment by CarterLi** [comment]
> Why can't tone mapping always be used?

I never say you can't. If the display doesn't support HDR mode, I just revert everything to default. It's not the best option, but at least doesn't make things worse.

At the time when I implemented HDR feature for IINA, there were a lot issues complaining that IINA's HDR mode was bad; its brightness didn't match QuickTime; someone even said that IINA's HDR mode made HDR video completely unwatchable and ask removing it, etc. Remember that? I actually didn't have time to pay attention to playing HDR video on SDR display, and to be honest, I burnt out

In addition, I don't think playing HDR video on SDR display is a good idea. That's it.

**Comment by low-batt** [comment]
@CarterLi That is what I was hoping the answer was. Being conservative back when this was added was definitely the correct choice. I remember all the problems in macOS that caused us trouble. I was checking with you to make sure there wasn't some mpv technical issue that we had to be concerned with that I didn't know about.

I'm about to be making changes related to color management as IINA needs to be updated to match up with mpv 0.38.0 (as much as possible). Will be starting on this soon.

**Comment by burgerga** [comment]
I'd be happy to assist implementing HDR to SDR tone-mapping when gpu-next is available. Is there somewhere you discuss features/implementation details?

**Comment by low-batt** [comment]
Thanks for the offer to assist! Contributions always welcome.

Discussing features/implementation details in GitHub issues is good as it provides a record and is available for others to read. You can also ask questions in the [Telegram Group](https://t.me/IINAUsers).

I've not been closely following `gpu-next` due to it not yet being available through the rendering API. If you browse to [mpv issues](https://github.com/mpv-player/mpv/issues) you will see a pinned "Want to work on mpv?" issue. That issue lists "add a way for libmpv to work with libplacebo/gpu-next", so the mpv project is asking for help with that task.

Once support is available there will be some significant work required in IINA to support `gpu-next`. I expect it will involve adding a Metal view layer.

Once that work is done then IINA settings such as the tone mapping algorithm will need to be updated to include new algorithms supported by `gpu-next`. Since `gpu-next` is currently still considered experimental if IINA was adding `gpu-next` support today it would need provide a choice to switch back to the old video driver. That would bring up the problem of how to deal with settings that configure features only supported by `gpu-next` when you switch back to the old driver. How would IINA deal with settings that are invalid after switching back to the old driver? That kind of user interface problem.
